### PR TITLE
FTS support

### PR DESCRIPTION
**Pull request by aayush3011**
## Purpose

* Adding FTS support.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Review by jcocchi — CHANGES_REQUESTED** [review]
I'm getting a bunch of build errors when I run this. Left the required fixes here.

File: infra/app/database.bicep [hunk 1]
```diff
@@ -60,6 +63,11 @@ var containers = [
           type: 'diskANN'
         }
       ]
+      fullTextIndexes: [
+        {
+          path: '/description',
```
Comment (jcocchi) [inline infra/app/database.bicep:68]: Weren't we going to add FTS search to the tags too?

File: infra/core/database/cosmos-db/account.bicep [hunk 2]
```diff
@@ -56,7 +56,14 @@ resource account 'Microsoft.DocumentDB/databaseAccounts@2024-05-15' = {
               name: 'EnableNoSQLVectorSearch'
             }
           ]
-        : []
+        : [],
+      (enableNoSQLFullTextSearch)
```
Comment (jcocchi) [inline infra/core/database/cosmos-db/account.bicep:60]: This parameter isn't defined anywhere and throws an error during bicep deployment. Need to add parameter definition at the top of this file.

Add this below line 15
```
@description('Enables NoSQL full text search for this account. Defaults to false.')
param enableNoSQLFullTextSearch bool = false
```

And we need to pass a value in from the `infra/app/database.bicep` file. Add the parameter below so starting on line 140 it should look like this
```
module cosmosDbAccount '../core/database/cosmos-db/nosql/account.bicep' = {
  name: 'cosmos-db-account'
  params: {
    name: accountName
    location: location
    tags: tags
    enableServerless: true
    enableVectorSearch: true
    enableNoSQLFullTextSearch: true // This is the new parameter we just defined
    disableKeyBasedAuth: true
  }
}
```

File: infra/app/database.bicep (hunk 1 above)
Comment (jcocchi) [inline infra/app/database.bicep:68]: Comma throwing syntax error

```suggestion
          path: '/description'
```

File: infra/app/database.bicep [hunk 3]
```diff
@@ -71,6 +79,15 @@ var containers = [
         }
       ]
     }
+    fullTextPolicy: {
+      defaultLanguage: 'en-US',
```
Comment (jcocchi) [inline infra/app/database.bicep:83]: Comma throwing syntax error

```suggestion
      defaultLanguage: 'en-US'
```

File: infra/app/database.bicep [hunk 4]
```diff
@@ -71,6 +79,15 @@ var containers = [
...
+      fullTextPaths: [
+        {
+          path: '/description',
```
Comment (jcocchi) [inline infra/app/database.bicep:86]: Comma throwing syntax error

```suggestion
          path: '/description'
```

File: infra/app/database.bicep [hunk 5]
```diff
@@ -92,6 +109,11 @@ var containers = [
           type: 'diskANN'
         }
       ]
+      fullTextIndexes: [
+        {
+          path: '/description',
```
Comment (jcocchi) [inline infra/app/database.bicep:114]: Comma throwing syntax error

```suggestion
          path: '/description'
```

File: infra/app/database.bicep [hunk 6]
```diff
@@ -103,6 +125,15 @@ var containers = [
         }
       ]
     }
+    fullTextPolicy: {
+      defaultLanguage: 'en-US',
```
Comment (jcocchi) [inline infra/app/database.bicep:129]: Comma throwing syntax error

```suggestion
      defaultLanguage: 'en-US'
```

File: infra/app/database.bicep [hunk 7]
```diff
@@ -103,6 +125,15 @@ var containers = [
...
+      fullTextPaths: [
+        {
+          path: '/description',
```
Comment (jcocchi) [inline infra/app/database.bicep:132]: Comma throwing syntax error

```suggestion
          path: '/description'
```

**Review by jcocchi — COMMENTED** [review]
Added more fixes for the bicep errors

File: infra/app/database.bicep (hunk 1 above)
Comment (aayush3011) [reply infra/app/database.bicep:68]: Updated to tags now.

File: infra/core/database/cosmos-db/account.bicep (hunk 2 above)
Comment (aayush3011) [reply infra/core/database/cosmos-db/account.bicep:60]: Added in the necessary places.

File: infra/app/database.bicep [hunk 8]
```diff
@@ -60,6 +63,11 @@ var containers = [
           type: 'diskANN'
         }
       ]
+      fullTextIndexes: [
+        {
+          path: '/tags'
```
Comment (jcocchi) [inline infra/app/database.bicep:68]: Should we leave the index on description too? We aren't sure what will end up working best so maybe leaving both description and tags gives us more properties to test the flow out on. It's hard to know for sure until we can test a few queries

Comment (aayush3011) [reply infra/app/database.bicep:68]: Added /description as well.

**Comment by aayush3011** [comment]
Closing this PR for now, will add FTS support to this lab post ignite.

**Comment by markjbrown** [comment]
@aayush3011 @jcocchi 

I am fixing things in this PR now. I updated the bicep and have deployed this successfully with the FTS index on products container. 

Right now I am updating the FTS query syntax. I'm stuck on an exception I'm getting with it complaining about an array of string literals. I've asked James to review the syntax.

One other change I'm making as well is to rewrite the token limit function to truncate products by item rather than strip characters in a string.

Once I get this FTS query resolved and update this token limit function I'll push into this PR and merge into aspire branch.

Thanks.

**Comment by markjbrown** [comment]
I'm trying to push the changes back into this PR but I am getting this error. I reauthorized VS but not working. Either of you know how to fix this?

Failed to push to the remote repository. See the Output window for more details.
Pushing aspire-pr-55
Remote: The 'AzureCosmosDB' organization has enabled or enforced SAML SSO.
Remote: To access this repository, you must re-authorize the OAuth Application 'GitHub CLI'.
Error encountered while pushing to the remote repository: Git failed with a fatal error.
Git failed with a fatal error.
unable to access 'https://github.com/AzureCosmosDB/cosmosdb-nosql-copilot.git/': The requested URL returned error: 403

**Pull request merged**